### PR TITLE
FLEX-2737 ~ Adds timestamp in EventNotificationRequest.

### PR DESCRIPTION
**Pull request by kevinsmeets**
- backwards compatible with older protocol versions
- tested with client applications

### CODE REVIEW

File: osgp-adapter-protocol-oslp-elster/src/main/java/com/alliander/osgp/adapter/protocol/oslp/elster/infra/networking/OslpChannelHandlerServer.java [hunk 1]
```diff
@@ -268,22 +264,16 @@ public void processSignedOslpEnvelope(final SignedOslpEnvelopeDto signedOslpEnve
...
-
-            try {
-                // Send the event notification to OSGP-CORE to save in the
-                // database.
-                this.deviceManagementService.addEventNotification(Base64.encodeBase64String(deviceId), event.getEvent()
-                        .name(), event.getDescription(), index);
-            } catch (final UnknownEntityException ex) {
-                LOGGER.error("handle event notification request exception", ex);
-                oslpStatus = Oslp.Status.REJECTED;
-            }
```
Comment (kevinsmeets) [inline osgp-adapter-protocol-oslp-elster/src/main/java/com/alliander/osgp/adapter/protocol/oslp/elster/infra/networking/OslpChannelHandlerServer.java]: The function addEventNotification() doesn't throw UnknownEntityException. Therefore, the try-catch was removed.

File: oslp/src/main/protobuf/oslp.proto [hunk 2]
```diff
@@ -365,6 +365,7 @@ message EventNotification {
     required Event event = 1;
     optional bytes index = 2; // [(nanopb).max_size=1];
     optional string description = 3; // [(nanopb).max_size = 81];
+    optional string timestamp = 4; // [(nanopb).max_size = 15]; // - Format YYYYMMDDhhmmss UTC, indicates the date and time of the event.
```
Comment (kevinsmeets) [inline oslp/src/main/protobuf/oslp.proto]: The new optional variable 'timestamp' which has been added to EventNotificationRequest.

File: web-device-simulator/src/main/java/com/alliander/osgp/webdevicesimulator/service/RegisterDevice.java [hunk 3]
```diff
@@ -380,7 +390,7 @@ public void checkSequenceNumber(final byte[] bytes, final Integer sequenceNumber
             throw new DeviceSimulatorException("sequence number Integer is null");
         }
 
-        final Integer num = ((bytes[0] & 0xFF) << 8) | ((bytes[1] & 0xFF));
```
Comment (kevinsmeets) [inline web-device-simulator/src/main/java/com/alliander/osgp/webdevicesimulator/service/RegisterDevice.java]: Removed unneeded parentheses.

File: web-device-simulator/src/main/java/com/alliander/osgp/webdevicesimulator/service/RegisterDevice.java [hunk 4]
```diff
@@ -94,252 +100,256 @@ public DeviceMessageStatus sendRegisterDeviceCommand(final long deviceId, final
...
-                        this.oslpPortClient), request, device.getDeviceIdentification());
-
-            } else if (device.getProtocol().equals(ProtocolType.OSLP_ELSTER.toString())) {
-                // Send registration message
-                response = this.oslpChannelHandler.send(new InetSocketAddress(this.oslpAddressServer,
-                        this.oslpElsterPortClient), request, device.getDeviceIdentification());
-            }
-
-            LOGGER.debug("Controller Received Send Event Notification Command: "
-                    + response.getPayloadMessage().toString());
```
Comment (kevinsmeets) [inline web-device-simulator/src/main/java/com/alliander/osgp/webdevicesimulator/service/RegisterDevice.java]: This logging statement has been replaced by the logging statements in the function sendRequest().

File: web-device-simulator/src/main/java/com/alliander/osgp/webdevicesimulator/service/RegisterDevice.java [hunk 5]
```diff
@@ -94,252 +100,256 @@ public DeviceMessageStatus sendRegisterDeviceCommand(final long deviceId, final
...
+        if (index == null) {
+            return 0;
+        } else {
+            return index;
+        }
+    }
+
+    private String formatTimestamp() {
+        final String format = "yyyyMMddHHmmss";
+        final DateTime dateTime = DateTime.now().toDateTime(DateTimeZone.UTC);
```
Comment (kevinsmeets) [inline web-device-simulator/src/main/java/com/alliander/osgp/webdevicesimulator/service/RegisterDevice.java]: The format and time zone are dictated by the OSLP specification.

File: web-device-simulator/src/main/java/com/alliander/osgp/webdevicesimulator/service/RegisterDevice.java [hunk 6]
```diff
@@ -94,252 +100,256 @@ public DeviceMessageStatus sendRegisterDeviceCommand(final long deviceId, final
...
+                                    .setConfirmRegisterDeviceRequest(
+                                            Oslp.ConfirmRegisterDeviceRequest.newBuilder()
+                                                    .setRandomDevice(device.getRandomDevice())
+                                                    .setRandomPlatform(device.getRandomPlatform())).build()).build();
+
+            // Write outgoing request to log.
+            this.writeOslpLogItem(oslpRequest, device, false);
 
-            LOGGER.debug("Controller Received Send Confirm Device Registration Command: "
-                    + response.getPayloadMessage().toString());
```
Comment (kevinsmeets) [inline web-device-simulator/src/main/java/com/alliander/osgp/webdevicesimulator/service/RegisterDevice.java]: This logging statement has been replaced by the logging statements in the function sendRequest().

File: web-device-simulator/src/main/java/com/alliander/osgp/webdevicesimulator/service/RegisterDevice.java [hunk 7]
```diff
@@ -94,252 +100,256 @@ public DeviceMessageStatus sendRegisterDeviceCommand(final long deviceId, final
...
-                response = this.oslpChannelHandler.send(new InetSocketAddress(this.oslpAddressServer,
-                        this.oslpElsterPortClient), olspRequest, device.getDeviceIdentification());
-            }
+                                                    .setHasSchedule(hasSchedule).setRandomDevice(randomDevice)).build())
+                    .build();
+
+            // Write outgoing request to log.
+            this.writeOslpLogItem(oslpRequest, device, false);
 
-            LOGGER.debug("Controller Received Send Register Device Command: " + response.getPayloadMessage().toString());
```
Comment (kevinsmeets) [inline web-device-simulator/src/main/java/com/alliander/osgp/webdevicesimulator/service/RegisterDevice.java]: This logging statement has been replaced by the logging statements in the function sendRequest().

File: osgp-adapter-protocol-oslp-elster/src/main/java/com/alliander/osgp/adapter/protocol/oslp/elster/infra/networking/OslpChannelHandlerServer.java [hunk 8]
```diff
@@ -182,11 +180,9 @@ public void processSignedOslpEnvelope(final SignedOslpEnvelopeDto signedOslpEnve
 
         final String deviceIdentification = registerRequest.getDeviceIdentification();
         InetAddress inetAddress = InetAddress.getByAddress(registerRequest.getIpAddress().toByteArray());
-        if (this.testDeviceId != null && this.testDeviceIp != null) {
-            if (deviceIdentification.equals(this.testDeviceId)) {
-                LOGGER.info("Using testDeviceId: {} and testDeviceIp: {}", this.testDeviceId, this.testDeviceIp);
-                inetAddress = InetAddress.getByName(this.testDeviceIp);
-            }
+        if (this.testDeviceId != null && this.testDeviceIp != null && deviceIdentification.equals(this.testDeviceId)) {
```
Comment (kevinsmeets) [inline osgp-adapter-protocol-oslp-elster/src/main/java/com/alliander/osgp/adapter/protocol/oslp/elster/infra/networking/OslpChannelHandlerServer.java]: Combined if statements in order to resolve SonarQube issue.

File: web-device-simulator/src/main/java/com/alliander/osgp/webdevicesimulator/service/RegisterDevice.java [hunk 9]
```diff
@@ -94,252 +100,256 @@ public DeviceMessageStatus sendRegisterDeviceCommand(final long deviceId, final
...
+            final Oslp.Event event, final String description, final Integer index, final String timestamp) {
+        final String deviceUid = device.getDeviceUid();
+
+        // Create an event notification depending on device protocol (for now
+        // with 1 event).
+        Oslp.EventNotification eventNotification = null;
+        if (device.getProtocol().equals(ProtocolType.OSLP.toString())) {
+            eventNotification = EventNotification.newBuilder().setEvent(event)
+                    .setDescription(description == null ? "" : description)
+                    .setIndex(ByteString.copyFrom(new byte[] { index.byteValue() })).build();
```
Comment (kevinsmeets) [inline web-device-simulator/src/main/java/com/alliander/osgp/webdevicesimulator/service/RegisterDevice.java]: This is the implementation for the original protocol.

File: web-device-simulator/src/main/java/com/alliander/osgp/webdevicesimulator/service/RegisterDevice.java [hunk 10]
```diff
@@ -94,252 +100,256 @@ public DeviceMessageStatus sendRegisterDeviceCommand(final long deviceId, final
...
+        } else if (device.getProtocol().equals(ProtocolType.OSLP_ELSTER.toString())) {
+            eventNotification = EventNotification.newBuilder().setEvent(event)
+                    .setDescription(description == null ? "" : description)
+                    .setIndex(ByteString.copyFrom(new byte[] { index.byteValue() })).setTimestamp(timestamp).build();
```
Comment (kevinsmeets) [inline web-device-simulator/src/main/java/com/alliander/osgp/webdevicesimulator/service/RegisterDevice.java]: This is the implementation with the new 'timestamp' variable for the newer version of the protocol.

File: web-device-simulator/src/main/java/com/alliander/osgp/webdevicesimulator/service/RegisterDevice.java [hunk 11]
```diff
@@ -94,252 +100,256 @@ public DeviceMessageStatus sendRegisterDeviceCommand(final long deviceId, final
...
+    }
+
+    private void writeOslpLogItem(final OslpEnvelope oslpEnvelope, final Device device, final boolean incoming) {
+        final OslpLogItem logItem = new OslpLogItem(oslpEnvelope.getDeviceId(), device.getDeviceIdentification(),
+                incoming, oslpEnvelope.getPayloadMessage());
+        this.oslpLogItemRepository.save(logItem);
+    }
+
+    private OslpEnvelope sendRequest(final Device device, final OslpEnvelope request) throws IOException,
+            DeviceSimulatorException {
```
Comment (kevinsmeets) [inline web-device-simulator/src/main/java/com/alliander/osgp/webdevicesimulator/service/RegisterDevice.java]: Combined the check on protocol in order to choose the correct port number and the sending of a request / receiving response in order to reduce duplicate code.

File: web-device-simulator/src/main/java/com/alliander/osgp/webdevicesimulator/service/RegisterDevice.java [hunk 12]
```diff
@@ -94,252 +100,256 @@ public DeviceMessageStatus sendRegisterDeviceCommand(final long deviceId, final
...
+    private int determineRelayIndexForEventNotification(final Integer index) {
+        // Test the index provided in request.
+        if (index == null) {
+            return 0;
+        } else {
+            return index;
+        }
+    }
+
+    private String formatTimestamp() {
```
Comment (kevinsmeets) [inline web-device-simulator/src/main/java/com/alliander/osgp/webdevicesimulator/service/RegisterDevice.java]: Renamed

File: osgp-adapter-protocol-oslp-elster/src/main/java/com/alliander/osgp/adapter/protocol/oslp/elster/application/services/DeviceManagementService.java [hunk 13]
```diff
@@ -53,25 +57,34 @@ public DeviceManagementService() {
...
      *            The index of the device.
+     * @param timestamp
+     *            The date and time of the event.
      */
     public void addEventNotification(final String deviceUid, final String eventType, final String description,
-            final Integer index) {
+            final int index, final String timestamp) {
+        Assert.notNull(eventType);
+        Assert.notNull(description);
+        Assert.notNull(timestamp);
```
Comment (kevinsmeets) [inline osgp-adapter-protocol-oslp-elster/src/main/java/com/alliander/osgp/adapter/protocol/oslp/elster/application/services/DeviceManagementService.java]: The argument 'timestamp' is no longer required. The value for 'timestamp' may be null or an empty string.

**Comment by kevinsmeets** [comment]
@rlemmers your review comments have been processed.

**Comment by rlemmers** [comment]
Looks okay to me.

**Pull request merged**